### ISSUE TITLE
Allow for injecting Directive0 before complete in Akka HTTP?

### ISSUE THREAD

**Issue by DWvanGeest** [twilio/guardrail#645]
Hello again!

In some circumstances, it may be very useful if the Guardrail user can inject a `Directive0` into the generated Akka HTTP routes. The user could pass this directive as an argument to `FooResource.routes` with a default argument of `pass`, which does nothing. Then this directive can be included in the generated routes immediately prior to `complete`.

My use case is that I have a `Directive0` which checks that the authenticated user has the correct OAuth scope to access a certain path. I don't want to run this check, or give an `AuthorizationFailedRejection`, unless everything else (path, HTTP method, etc.) matches. 

Currently, I do wrap the generated route in my directive (similar to [this suggestion](https://github.com/twilio/guardrail/issues/90#issuecomment-413899877), which is OK-ish, but gives rise to confusing response codes. If my scope-protected route is `GET /hello` and someone tries `GET /bonjour` without the required scope, they'll get `403 Unauthorized` back. It's technically correct, but can lead the client to conclude that the `GET /bonjour` path actually exists. Essentially this is very similar to my issue #643 which gets at the precedence of HTTP response codes, and 404 having higher precedence than 403.

Another (theoretical) example of how this could be used is for metrics. You could inject a `Directive0` which increments a request counter for a specific path. I'm sure there are other ways to use such an escape hatch.

I'd be willing to contribute this. Do you think it would be a good idea?

**Comment by kelnos** [comment]
For your specific use case, we should implement the OAS security goo (https://swagger.io/docs/specification/authentication/), which should support everything you need.  It specifically supports OAuth2 and specifying scopes.

I did some work last year to expose the security stuff in the generators.  Handling the semantics of the various combinations they can be expressed (it's honestly a little _too_ flexible for my taste) turns out to be kinda difficult; I did some work on a Dropwizard implementation, but never finished it.  It's possible that akka-http would be easier to do.

**Comment by DWvanGeest** [comment]
Hey @kelnos, thanks for your response. 

While I'm generally in agreement that it would be good to implement the OAS security stuff, I do fear that it would be quite tricky to make a good abstraction, as you perhaps experienced. One thing that comes to mind at the moment: I'm wondering how the scopes of the authenticated user would be determined. In my particular case they can be extracted from a signed JWT, but I don't think that's a part of either the [OAuth 2](https://tools.ietf.org/html/rfc6749#section-3.3) or OAS standards. Other applications could, for example, verify an `access_token` by calling an external service, which returns the granted scopes in its response. I'm sure there are plenty of other similar questions.

OAuth2 scopes aside, I'm seeing this as a more broad way to customize the generated routes. I think there are other reasons besides OAuth that users may want to inject behavior between the generated route and the handler. 

To rephrase the question: how can guardrail allow the user to implement server behavior that is not covered by the OAS spec, but is also not specific to a handler implementation? I think allowing for a `Directive` hook between the spec-generated route and the user-defined handler would make things a lot more flexible. It would cover the standard auth cases, but it would also cover other things. I'm thinking of it as an "escape hatch" similar in philosophy to `x-scala-type` which we've found to be quite useful.

**Comment by kelnos** [comment]
I get what you're saying, but I think we should implement the spec to cover the use cases we want, before we implement custom behavior.  Understand that your use case might not be completely covered by the spec and you'd need something custom anyway.  I'm not opposed to this kind of thing, but I'll leave it up to @blast-hardcheese to muse about what that could look like.

**Comment by blast-hardcheese** [comment]
Thank you for putting this together!

I'm concerned that this doesn't go far enough. guardrail has a solid history of baking in escape hatches to support functionality that is not able to be easily expressed in OpenAPI, as well as to work around bugs in guardrail, so this isn't a huge concern of mine.

What is a concern is minimizing API changes on upgrade. We've had the benefit of a very few changes to guardrail that require altering consumer code on upgrade, even though we haven't hit 1.0 this is a pretty significant design goal of mine.

Given this, I think finding a solid encoding for holding already-extracted parameters, describing as much data as we have at that moment, and passing _that_ as a parameter into `beforeComplete`, instead of just the method name. Stringy matching of `operationId` creates an unnecessarily brittle function, and also opens us up to requests to add more context going forward, which will cause a neverending cycle of breaking source compatibility for anyone using this feature.

Unfortunately, this drastically expands the scope of the feature, and muddies the water of the difference between `beforeComplete` and each individual `Handler` function, as all routing could now just be done in a giant `match` statement on each routing member of a sealed trait, which gets closer to how routing works in http4s land (and makes domain routing logic just another function, instead of implementing an interface, which has even more implications for compositionality).

All in all, despite the work you've done in #656 and the clear need for solving the problem you're facing, I think this warrants some exploration into the ergonomics and utility of an approach like this.

If you're up to it, I'd request generating your servers with guardrail from your specification, then moving the generated code into your `src` and modifying it to match what I've described here. For example,

```scala
trait Handler {
  def postTest(respond: Resource.postTestResponse.type)(body: TestObject): Future[Resource.postTestResponse]
}
```

would be augmented with
```scala
sealed trait Requests
case class PostTestRequest(body: TestObject) extends Requests
```

or even
```scala
sealed trait Requests[Response]
case class PostTestRequest(body: TestObject) extends Requests[Resource.postTestResponse]
```
to permit a `FunctionK[Requests, Future]` encoding...

then, to expose `respond` maybe even something along the lines of
```scala
sealed trait Requests[Response]
case class PostTestRequest(respond: Resource.postTestResponse.type)(body: TestObject) extends Requests[Resource.postTestResponse]
```
in order to expose the `respond` companion conveniently.

I think this solves a more broad problem as well, where the akka-http generators are hard-coded to `Future` in all responses, making it difficult to generate `Arbitrary` instances for `Handler` methods, which makes writing property tests extraordinarily painful.

This could end up permitting a `compile` step of `F ~> Future`, permitting `EitherT` or ZIO or some other `F` to be translated into a supported response type wrapped in `Future`.

One step at a time though, just trying to think ahead a little as far as the gains this strategy buys us as far as quality, hygiene, and not compromising on the core value of generating high-quality code from OpenAPI specifications.

The nice thing about everything I've described here is that the adapter from what's described here and the existing `Handler` is a no-op, so the transition need not even be done, should the user not actually want or need that functionality (not everybody is interested in `FunctionK`, as it turns out).

**Comment by DWvanGeest** [comment]
Thanks for sharing your thoughts @blast-hardcheese, it's helpful to hear about your goals and philosophy on this kind of thing.

Your concern about source compatibility makes sense, and it does seem kind of weird to do all the param/body extraction before `beforeComplete`, only to not have it available. Finding a generic encoding is a good idea from that perspective, even if it's a lot more scope to take on.

But I also agree that adding that kind of feature would muddy the water between a handler and `beforeComplete`. And while this sort of `FunctionK` interface might be the norm for http4s, it's a bit far from Akka HTTP in my mind, and I doubt I would end up using it to anything near its full potential, personally.

All of this together makes me wonder if my proposed feature simply needs to be moved to a better spot. The sorts of use cases I'm envisioning (auth, monitoring, etc) don't typically require access to params or the deserialized body. Having only the path and method matched before running custom behavior would be sufficient and probably even preferable (e.g. the server shouldn't bother doing intensive deserialization if authorization is going to fail anyway).

What if I changed `beforeComplete` to be `afterPathMethodMatch: String => Directive0` and inserted it immediately after path and method matching? The only "decoding" work the Guardrail code has done is figuring out the `operationId`. Without any additional context decoded, nobody will be asking for additional parameters to this function (and they can use regular Directives to extract things from the request if needed). Would this address your source compatibility concern?

I don't think this precludes also doing the `FunctionK` stuff. But I will say that my capacity to take on that larger chunk of work is pretty limited. I'm following your general idea, but the implementation would be tricky for me.

To summarize, I think what I'm saying is that my feature fell into a weird middle-ground between simplicity and power. You correctly called out that it was neither, so I would propose intentionally moving it to the side of simplicity, and leaving the power feature for another day.

**Comment by DWvanGeest** [comment]
For the sake of demonstration/argument, I implemented my revised idea from the above comment in https://github.com/twilio/guardrail/pull/656. Let me know what you think!

**Comment by DWvanGeest** [comment]
I've been thinking about this some more and reading some other issues. Going to try a different approach which may be more general.

**Comment by DWvanGeest** [comment]
Fixed in #747
